docs/specs/user-model.md

```diff
@@ -1,71 +1,127 @@
+
+## Custom Attributes
+
+Custom Attributes is a JSON Object. The developer can update it via Admin API.
+
+Custom Attributes can be any valid JSON Object. There is no limitation on the nesting level. However, a sensible size limit (e.g. 10MiB) is enforced.
+
+The developer can save whatever they want in Custom Attributes.
+
+The user cannot read or write Custom Attributes directly.
+
+> NOTE: The initial implementation is very likely to be a single textarea for editing the whole custom attributes object.
+> Much more fancier features like generated form fields requires us to interpret the JSON schema and generate form fields.
+> That could be a very useful standalone product.
+
+### Custom Attributes validation
+
+The validation on custom attributes can be done with the JSON schema provided by the developer in the configuration.
+
+The JSON schema must be written against [the version 2019-09](https://json-schema.org/specification-links.html#2019-09-formerly-known-as-draft-8).
+
+For example,
+
+```yaml
+custom_attributes:
+  json_schema:
+    type: object
+    properties:
+      email:
+        type: string
+        format: email
+    required: ["email"]
+```
+
+Any update on custom attributes must be valid against the given schema.
+
+> NOTE: In the future, we can add our own JSON schema keyword to annotate the access control of individual fields in the custom attributes.
 
 ## Identity
 
 An identity is used to look up a user.
 
 3 types of identity are supported.
 
-- Login ID
-- OAuth
-- Anonymous
+- Login ID (`login_id`)
+- OAuth (`oauth`)
+- Anonymous (`anonymous`)
 
 ### Identity Claims
 
-The information of an identity are mapped to [Standard Claims](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims)
+- OAuth identity has `email` claim.
+- Email Login ID identity has `email` claim.
+- Phone Login ID identity has `phone_number` claim.
+- Username Login ID identity has `preferred_username` claim.
 
-Currently, only `email` is mapped.
+> TODO: In the future, we want to allow the developer to customize the scope when we perform OAuth flow with external OAuth provider.
+> Then we can support more Standard Claims, such as `picture`.
 
-The claims are used to detect duplicate identity. For example, an Email Login ID and the email claim of an OAuth Identity. This prevents duplicate user when the user forgets the original authentication method.
+The claim `email` is used to detect duplicate identity. For example, an Email Login ID and the email claim of an OAuth Identity. This prevents duplicate user when the user forgets the original authentication method.
 
 ### OAuth Identity
 
 OAuth identity is external identity from supported OAuth 2 IdPs. Only authorization code flow is supported. If the provider supports OIDC, OIDC is preferred over provider-specific OAuth 2 protocol.
 
+OAuth identity is updated every time authentication is performed.
```

Comment:
If an OAuth identity is updated to have the same email as an existing login ID identity in system, it would result in a conflict. This is related to the merge flow in #261.

I think we can clarify the for now, if updating an OAuth identity would result in a conflict, the update is skipped?